### PR TITLE
Issues 856 and 656: Optional parens in "CREATE TABLE AS", "optional bracket" parser support

### LINKED ISSUE #856: [BigQuery]  Couldn't find closing bracket for opening bracket.
Hi there!

I am running sqlfluff on a BigQuery update query and the linter throws a `Couldn't find closing bracket for opening bracket.
` error.

## lint 

```sql
CREATE TEMPORARY TABLE final_rows AS
(
    WITH source_table AS
    (
        SELECT
        sd.*,
        ccdv.version_id
        FROM
        source_data AS sd
        JOIN current_carrier AS ccdv
        ON ccdv.origin_country_id=sd.origin_country_id
        WHERE
        ccdv.destination_country_id=sd.destination_country_id
        AND ccdv.carrier_id=sd.carrier_id
        AND sd.order_date < sd.ship_date
        AND sd.ship_date < sd.delivery_date
    )
    SELECT
    sdp.po_number,
    cspo.station AS origin_station,
    cspd.station AS destination_station
    FROM
    source_table AS sdp
    LEFT JOIN `another_table` AS cspo
    ON sdp.origin=cspo.id
    LEFT JOIN `yet_another_table` AS cspd
    ON sdp.destination_country_id=cspd.id
    WHERE
    sdp.carrier_id=cspo.id
    AND sdp.carrier=cspo.version
    AND sdp.carrier_id=cspd.id
)
;
```

## Expected Behaviour

No errors

## Observed Behaviour
```
L:   4 | P:   5 |  PRS | Couldn't find closing bracket for opening bracket.

```

## Steps to Reproduce

```
sqlfluff lint file_name
```

## Version
Include the output of `sqlfluff --version` along with your Python version

## Configuration
```
[sqlfluff]
verbose = 0
dialect = bigquery
templater = python
```

### PR DESCRIPTION
**Pull request by barrywhart**
Resolves #856, #650, #693

This PR fixes an ANSI grammar bug: `CREATE TABLE AS` should allow but not require parentheses.

It also implements an `optionally_bracketed` function, as described in #650 and #693. It uses the new function, but does not update pre-existing code which would benefit from this function.

I debated whether to implement `optionally_bracketed` as a **class** inheriting from `OneOf`. This would certainly work. I went with the function version, since the implementation is a bit simpler (avoids inheritance that seems questionable because it's not a clean "is-a" relationship), more like macro expansion (?). This does affect the name, though, since PEP-8 uses a different naming convention for classes vs functions. Thoughts are welcome...

### CODE REVIEW

**Comment by NiallRees** [comment]
> Resolves #856, #650, #693
> 
> This PR fixes an ANSI grammar bug: `CREATE TABLE AS` should allow but not require parentheses.
> 
> It also implements an `optionally_bracketed` function, as described in #650 and #693. It uses the new function, but does not update pre-existing code which would benefit from this function.
> 
> I debated whether to implement `optionally_bracketed` as a **class** inheriting from `OneOf`. This would certainly work. I went with the function version, since the implementation is a bit simpler (avoids inheritance that seems questionable because it's not a clean "is-a" relationship), more like macro expansion (?). This does affect the name, though, since PEP-8 uses a different naming convention for classes vs functions. Thoughts are welcome...

Looks like you went for a class instead of a function in the end? How come OOI?

File: src/sqlfluff/core/parser/grammar/anyof.py [hunk 1]
```diff
@@ -221,3 +221,10 @@ class OneOf(AnyNumberOf):
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, max_times=1, min_times=1, **kwargs)
+
+
+class OptionallyBracketed(OneOf):
```
Comment (NiallRees) [inline src/sqlfluff/core/parser/grammar/anyof.py:226]: Nice! This will let us tidy up lots of places

**Review by NiallRees — APPROVED** [review]
🛳

**Comment by barrywhart** [comment]
I went with the class because that was consistent with other segments. TBH, the bigger reason is that if you follow PEP-8, the naming style for functions versus classes is so different that using a mix of both for defining a dialect was jarring. I admit it's kind of a subjective reason, and I don't feel strongly about it.

**Comment by NiallRees** [comment]
> I went with the class because that was consistent with other segments. TBH, the bigger reason is that if you follow PEP-8, the naming style for functions versus classes is so different that using a mix of both for defining a dialect was jarring. I admit it's kind of a subjective reason, and I don't feel strongly about it.

With you the whole way!

**Comment by barrywhart** [comment]
`flake8` is failing in code that is not touched by this PR. Any ideas, @NiallRees?

**Pull request merged**